### ISSUE TITLE
Generic OAuth: Confusing Documentation

### ISSUE THREAD

**Issue by russorat** [influxdata/docs.influxdata.com#1384]
_From @jcmcken on October 3, 2017 19:19_

The generic OAuth provider documentation states:

```text
The generic OAuth2 provider is very similar to the Github provider, but, you are able to set your own authentication, token, and API URLs. The callback URL path will be /oauth/generic/callback. So, if your Chronograf is hosted at https://localhost:8888 then the full callback URL would be https://localhost:8888/oauth/generic/callback.
```

This isn't actually true if you set ``GENERIC_NAME``, in which case the callback URL includes the generic name in the URL ([cf](https://github.com/influxdata/chronograf/blob/8a0498162817b1850d05fe5c2a9032828a10b98e/server/server.go#L218)).

_Copied from original issue: influxdata/chronograf#2065_

**Comment by russorat** [comment]
_From @stevebang on January 2, 2018 23:37_

Fixing this in the Chrono 1.4 docs -- unfortunately, this also frustrated @trucaliber in https://github.com/influxdata/docs.influxdata.com/issues/1365.

**Comment by russorat** [comment]
_From @stevebang on January 3, 2018 0:7_

Also, found that @austinbutler encountered this issue, based on his posting in the InfluxData community (https://community.influxdata.com/t/oauth-with-internal-only-domain/1376).

**Comment by russorat** [comment]
connect #2509 #2677

**Comment by jaredscheib** [comment]
@russorat I believe you meant:
connect https://github.com/influxdata/chronograf/issues/2509 https://github.com/influxdata/chronograf/issues/2677

**Comment by jaredscheib** [comment]
Missing from documentation: You must provide a PUBLIC_URL (e.g. `http://localhost:8888`). Generic constructs the redirect URL for you from that.

**Comment by jaredscheib** [comment]
Also note that Heroku does not work with Generic OAuth currently, but a fix for it working as a specific OAuth provider has been submitted at https://github.com/influxdata/chronograf/pull/2911.

**Comment by jaredscheib** [comment]
Also note that all generic callback URLs are constructed as `<PUBLIC_URL>/oauth/<GENERIC_NAME/callback`.

**Comment by jaredscheib** [comment]
Also please note that scopes should be `space-delimited`, not `comma-delimited`, such as (for a GitHub example) `GENERIC_SCOPES="user:email read:org".

**Comment by jaredscheib** [comment]
The the docs section `Creating OAuth 2.0 applications using your own provider`, it states that `GENERIC_SCOPES` and `GENERIC_API_URL` are _optional_, but this is not the case as sometimes they are required for certain Providers. Additionally, `PUBLIC_URL` is not even listed as a configuration option, but it is also required by some Providers.

"Optional" is probably not the best word -- it should actually communicate that some of these settings may be required for various Providers, rather than that they are optional (which sounds more like "optional to the user").

Okta, for example, requires the above listed three in order to authenticate successfully. Google does as well.

**Comment by jaredscheib** [comment]
Also, it should be noted that `user:email` may not be valid for many Providers, even if it is a default value.

**Comment by jaredscheib** [comment]
Finally, for Generic, the name of the Provider to be used when creating a user, or when creating a mapping, should be `generic` or whatever value was used for `GENERIC_NAME`.

**Comment by jaredscheib** [comment]
Closed by 2a87d8185d09b23031b68d59caede189d25d01b9
